### PR TITLE
[SMALLFIX] [s3] Add S3 DELETE bucket API

### PR DESCRIPTION
**Pull request by ifcharming**
TODO: add integration tests.

### CODE REVIEW

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java [hunk 1]
```diff
@@ -23,8 +23,10 @@
    */
   public static final class Name {
     public static final String BUCKET_ALREADY_EXISTS = "BucketAlreadyExists";
+    public static final String BUCKET_NOT_EMPTY = "BucketNotEmpty";
     public static final String INTERNAL_ERROR = "InternalError";
     public static final String INVALID_BUCKET_NAME = "InvalidBucketName";
+    public static final String NO_SUCH_BUCKET = "BucketNotExists";
```
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: why not be consistent as `BUCKET_NOT_EXISTS`?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java [hunk 2]
```diff
@@ -45,6 +51,10 @@ private Name() {
       Name.INTERNAL_ERROR,
       "We encountered an internal error. Please try again.",
       Response.Status.INTERNAL_SERVER_ERROR);
+  public static final S3ErrorCode NO_SUCH_BUCKET = new S3ErrorCode(
```
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: ditto

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 3]
```diff
@@ -106,13 +111,68 @@ public Void call() throws S3Exception {
     });
   }
 
+  /**
+   * @summary deletes a bucket
+   * @param bucket the Alluxio path
```
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: the bucket name

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 4]
```diff
@@ -106,13 +111,68 @@ public Void call() throws S3Exception {
...
+   */
+  @DELETE
+  @Path(BUCKET_PARAM)
+  @ReturnType("java.lang.Void")
+  public Response deleteBucket(@PathParam("bucket") final String bucket) {
+    return S3RestUtils.call(bucket, new S3RestUtils.RestCallable<Void>() {
+      @Override
+      public Void call() throws S3Exception {
+        String bucketPath =  AlluxioURI.SEPARATOR + bucket;
+        if (bucketPath.contains(BUCKET_SEPARATOR)) {
```
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: this block is identical to that in createBucket, can be put into a helper method

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java [hunk 5]
```diff
@@ -106,13 +111,68 @@ public Void call() throws S3Exception {
...
+            throw new InvalidPathException("Bucket name is not a valid Alluxio directory.");
+          }
+        } catch (Exception e) {
+          throw toBucketS3Exception(e, bucketPath);
+        }
+
+        // Delete the bucket.
+        DeleteOptions options = DeleteOptions.defaults();
+        options.setAlluxioOnly(Configuration.get(PropertyKey.PROXY_S3_DELETE_TYPE)
+            .equals(S3RestUtils.DELETE_IN_ALLUXIO_ONLY));
```
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: would it be better to put `DELETE_IN_ALLUXIO_ONLY` into `alluxio.Constants`?

File: core/common/src/main/java/alluxio/PropertyKey.java [hunk 6]
```diff
@@ -1181,6 +1181,14 @@ public String toString() {
               + "`CACHE_THROUGH` (try to cache, write to UnderFS synchronously), "
               + "`THROUGH` (no cache, write to UnderFS synchronously).")
           .build();
+  public static final PropertyKey PROXY_S3_DELETE_TYPE =
+      new Builder(Name.PROXY_S3_DELETE_TYPE)
+          .setDefaultValue("ALLUXIO_AND_UFS")
```
Comment (uronce-cc) [inline core/common/src/main/java/alluxio/PropertyKey.java]: put `ALLUXIO_AND_UFS` into `alluxio.Constants`?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java (hunk 1 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: I'm following the spec here: http://docs.aws.amazon.com/AmazonS3/latest/API/ErrorResponses.html

It's called `NoSuchBucket` in s3 error code docs.

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java (hunk 2 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: same

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java (hunk 1 above)
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: then the string should be "NoSuchBucket"?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 3 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 4 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done, thanks

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java (hunk 5 above)
Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3RestServiceHandler.java]: done

File: core/common/src/main/java/alluxio/PropertyKey.java (hunk 6 above)
Comment (ifcharming) [inline core/common/src/main/java/alluxio/PropertyKey.java]: done

**Comment by ifcharming** [comment]
@uronce-cc I've addressed your comments. Can you PTAL?

File: core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java (hunk 1 above)
Comment (uronce-cc) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: @ifcharming

Comment (ifcharming) [inline core/server/proxy/src/main/java/alluxio/proxy/s3/S3ErrorCode.java]: oops, forgot to update this one, done

**Comment by ifcharming** [comment]
@calvinjia Please do a post-merge review. I'll address any comments afterwards.
Since it's targeting on a feature branch, I'd like to merge it once build passes.

**Comment by ifcharming** [comment]
@uronce-cc I've added some tests for DELETE bucket. PTAL

**Pull request merged**